### PR TITLE
Add ID token validation

### PR DESCRIPTION
**Pull request by joshcanhelp**
This PR adds ID token validation to the basic OmniAuth-Auth0 strategy. The main strategy was not altered in any way functionally except to add this validation (minor docs additions and minor formatting). The main work for this PR is in the new `OmniAuth::Auth0::JWTValidator` class. 

This uses the [Ruby JWT library](https://github.com/jwt/ruby-jwt) and validates the following:

- Signature
- Expiration (30 sec leeway)
- Issuer
- Audience
- Not Before (30 sec leeway)

### CODE REVIEW

**Comment by esarafianou** [comment]
LGTM for a security perspective

File: lib/omniauth/auth0/jwt_validator.rb [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+        # Docs: https://github.com/jwt/ruby-jwt#algorithms-and-usage
+        JWT.decode(jwt, decode_key, true, decode_options)
+      end
+
+      # Get the decoded head segment from a JWT.
+      # @return hash - The parsed head of the JWT passed, empty hash if not.
+      def token_head(jwt)
+        jwt_parts = jwt.split('.')
+        return {} if blank?(jwt_parts) || blank?(jwt_parts[0])
```
[the remark this answers is not in the document]
Comment (joshcanhelp) [reply lib/omniauth/auth0/jwt_validator.rb]: > If you return {} from here, can the rest of the library execute?

The part of this lib that uses it will fail on looking for an algorithm, which seems rational to me for an empty head. 

> What would you name this line if you had to give it a name?

Not sure what you're asking here. 

> What are we checking for?

Whether or not was have something to parse.

File: lib/omniauth/auth0/jwt_validator.rb [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+        # https://docs.ruby-lang.org/en/2.4.0/OpenSSL/X509/Certificate.html
+        OpenSSL::X509::Certificate.new(x5c).public_key
+      end
+
+      # Generate a URL with a domain name.
+      # @return string - The @domain property as a URL.
+      def domain_url
+        domain_url = URI(@domain)
+        domain_url = URI("https://#{@domain}") if domain_url.scheme.nil?
```
[the remark this answers is not in the document]
Comment (joshcanhelp) [reply lib/omniauth/auth0/jwt_validator.rb]: Can you rephrase? This allows a URI to be loaded without `http://` or `https://` and will convert it.

File: lib/omniauth/auth0/jwt_validator.rb [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+      end
+
+      # Return a specific key from a JWKS object.
+      # @param key string - Key to find in the JWKS.
+      # @param kid string - Key ID to identify the right JWK.
+      # @return nil|string
+      def jwks_key(key, kid)
+        @jwks ||= fetch_jwks
+        return nil if blank?(@jwks[:keys])
+        @jwks[:keys].each do |jwk|
```
[the remark this answers is not in the document]
Comment (joshcanhelp) [reply lib/omniauth/auth0/jwt_validator.rb]: I agree RE: where this is memoized so I moved that to `fetch_jwks`. But I think you're misunderstanding what this is doing. The JWKS is like:

```json
{
  keys [
    {
      "kid" : "kid_1",
      "key" : "thing_we_want"
    }, {
      "kid" : "kid_2",
      "key" : "thing_we_want"
    }
  ]
}
```

... in the general case so we need to iterate through the keys, check the `kid` for each, and return the value if we have a matching `kid`. Otherwise return nil. I simplified it a bit based on your feedback but what you're proposing won't do what we need here.

File: lib/omniauth/auth0/jwt_validator.rb [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+          next if jwk[:kid] != kid
+          return jwk[key]
+        end
+        nil
+      end
+
+      # Get a JWKS from the issuer
+      # @param path string - JWKS path from issuer.
+      # @return void
+      def fetch_jwks(path = '.well-known/jwks.json')
```
[the remark this answers is not in the document]
Comment (joshcanhelp) [reply lib/omniauth/auth0/jwt_validator.rb]: It was but I think it should be private in this case.

File: lib/omniauth/auth0/jwt_validator.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (machuga) [reply lib/omniauth/auth0/jwt_validator.rb]: I'm mostly just being picky about the Ruby idioms in this situation!

File: lib/omniauth/auth0/jwt_validator.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (machuga) [reply lib/omniauth/auth0/jwt_validator.rb]: Sure! So here you're assigning to `domain_url` twice (you're also shadowing the method name itself which is confusing) and I get why, it just feels odd. However, I think moving this up to the constructor would make sense anyway. For instance: 

```ruby
def initialize(options)
        @issuer = "#{domain_url}/"
        @client_id = options.client_id
        @client_secret = options.client_secret
        @jwks = nil

        temp_domain = URI(options.domain)
        @domain = temp_domain.scheme? ? temp_domain : URI("https://#{options.domain}")
      end
```
Then this method can simply return `@domain.to_s`

[the remark this answers is not in the document]
Comment (joshcanhelp) [reply lib/omniauth/auth0/jwt_validator.rb]: @machuga - Good comment. I ended up removing this method and the `@domain` property since it was only used for `@issuer`.
